port library to AnyIO

ioistired opened the pull request:
This allows users to manipulate stickers in an async (asyncio, curio, or trio) context, such as a web server or Signal bot.
It also enables us to download the pack images in parallel.

ioistired committed:
port library to AnyIO

ioistired committed:
relax asks requirement to allow installing from pypi

ioistired commented:
Marked as draft until I convert the library to use `httpx` instead of `asks`.

ioistired committed:
port to httpx

this allows us to use the latest anyio

ioistired committed:
only upload up to 5 stickers in parallel to prevent http issues

ioistired committed:
add higher timeouts to all http calls

ioistired committed:
re-use the http session object

ioistired committed:
30s timeout still wasn't enough

ioistired committed:
add some GET request status code validation

romainricard commented on tests.py:
```diff
@@ -1,39 +1,36 @@
-class TestSignalStickersClient(unittest.TestCase):
+# This is the same as using the @pytest.mark.anyio on all test functions in the module
+pytestmark = pytest.mark.anyio
 
```
My `pytest` tries to use `trio` instead of `asyncio`. It could be useful to specify a backend:

```python
@pytest.fixture
def anyio_backend():
    return 'asyncio'
```

ioistired replied:
yes. it tires to test on asyncio, trio, and curio. i left it that way to ensure that everything works on all three platforms. i will ensure the CI does so as well, unless you have a specific reason to only test asyncio?

ioistired replied:
oh there is no CI, is there

romainricard replied:
Yep, there is no CI on this project (yet). The main reason is that `asyncio` is a Python standard library, where `curio` and `trio` have to be installed, which can be annoying if you just want to contribute a small fix and run tests before sending. However you have a point in willing to test on all three

romainricard commented on examples/upload_stickers.py:
```diff
@@ -1,49 +1,55 @@
-print("Pack uploaded!\n\nhttps://signal.art/addstickers/#pack_id={}&pack_key={}".format(pack_id, pack_key))
\ No newline at end of file
+if __name__ == '__main__':
+    import anyio
```
Put this at the top :top:

romainricard reviewed (changes requested):
Thank you very much for this PR! Performances are greatly enhanced :partying_face: 
I've just two minors remarks. Other than that, it's perfect!

ioistired committed:
incorporate suggestions from review

- move an import in examples/upload_stickers.py to the top of the file
- include curio and trio in the dev dependencies for testing

ioistired commented:
Should be good now.

romainricard commented:
Thanks!

ioistired commented:
Phew, feels good :shipit: :sweat_smile:
